Video download button does not appear in course (TNL-5769)

noraiz-anwar opened the pull request:
This PR is made in effort to resolve [TNL-5769](https://openedx.atlassian.net/browse/TNL-5769)

A brief summary and attempted solution is as follows:

- Despite having the Video Download Allowed set to True, the download button was not appearing for the video.
- Video ID was extended by '\t'(tab). This extra tab was causing a problem while fetching video by edxval api so download link was not rendering.
- The problem is resolved by stripping Video ID before sending it to edxval api method which fetches video from the database.
- A test is written to ensure any valid Video ID having unwanted space(s) before or after it should not cause any problem.

###  Sandbox
[Sandbox](https://studio-tnl-5769.sandbox.edx.org) is created to test this implementation.
### How to test?
Goto sandbox and edit any course and add a video component with the following properties
**Default Video URL**: _https://www.youtube.com/watch?v=-8d9jHTWhw0_
**Video Download Allowed**: _true_    
...

noraiz-anwar commented:
jenkins run quality

noraiz-anwar commented:
jenkins run lettuce

noraiz-anwar commented:
jenkins run quality

noraiz-anwar commented:
jenkins run bokchoy

noraiz-anwar commented:
@awaisdar001 , @Ayub-Khan please review.

noraiz-anwar commented:
jenkins run quality

awaisdar001 commented on lms/djangoapps/courseware/tests/test_video_mongo.py:
```diff
@@ -582,6 +582,7 @@ def test_get_html_with_mocked_edx_video_id(self):
 
+    # pylint: disable=invalid-name
     def test_get_html_with_existing_edx_video_id(self):
         # create test profiles and their encodings
```
^^ Doc-string

awaisdar001 commented:
@noraiz-anwar Have you wrote a new test for this? Seems like you have editing an existing one.

noraiz-anwar commented:
yes I just added a new case in an existing test. That test was checking video by an existing edx_video_id. I added in test so that an edx_video_id with extra tab can be tested.

awaisdar001 commented:
@noraiz-anwar Don't you think testing video with a tab based `video_id` yields a new scenrio?

noraiz-anwar commented:
Yes @awaisdar001 this is a new scenario.

awaisdar001 commented:
How do you think testing this in a `new` & `seperate` test?

noraiz-anwar commented:
@awaisdar001 New and separate test would be a lot of repetition. Case we are trying to test is getting a video, using 'Video ID plus some extra space'. The already written test is doing all the same things except video ID is fair i.e no extra spaces. I tried to use same code but an added case where video ID contained an extra tab.

awaisdar001 commented:
Now thats' the question i would ask, How we can make it a seperate test and not repeat the code? :)

noraiz-anwar commented:
@awaisdar001 I am looking at that. Hope I'll be able to answer this soon.

awaisdar001 commented:
Perfect.

awaisdar001 commented on common/lib/xmodule/xmodule/video_module/video_module.py:
```diff
@@ -217,7 +217,7 @@ def get_html(self):
             try:
                 val_profiles = ["youtube", "desktop_webm", "desktop_mp4"]
-                val_video_urls = edxval_api.get_urls_for_profiles(self.edx_video_id, val_profiles)
+                val_video_urls = edxval_api.get_urls_for_profiles(self.edx_video_id.strip(), val_profiles)
```
Please add a comment why we are striping the `self.edx_video_id`

awaisdar001 commented on lms/djangoapps/courseware/tests/test_video_mongo.py:
```diff
@@ -607,33 +653,19 @@ def test_get_html_with_existing_edx_video_id(self):
-                           [video['url'] for video in encoded_videos],
-            }
-        }
+                    <video show_captions="true"
```
Maybe an extra tab is added for identation?

noraiz-anwar commented:
jenkins run quality

noraiz-anwar commented:
jenkins run python

noraiz-anwar commented:
jenkins run bokchoy

awaisdar001 commented on common/lib/xmodule/xmodule/video_module/video_module.py:
```diff
@@ -217,7 +217,9 @@ def get_html(self):
                 val_profiles = ["youtube", "desktop_webm", "desktop_mp4"]
-                val_video_urls = edxval_api.get_urls_for_profiles(self.edx_video_id, val_profiles)
+
+                # strip edx_video_id to prevent ValVideoNotFoundError error if unwanted spaces are there.
```
Please include JIRA# in the comment.

awaisdar001 commented on lms/djangoapps/courseware/tests/test_video_mongo.py:
```diff
@@ -583,30 +583,89 @@ def test_get_html_with_mocked_edx_video_id(self):
+        Tests the VideoModule get_html where a edx_video_id is given and related video is found
+        """
+        edx_video_id = 'thundercats'
+        encoded_videos = self.encode_and_create_video(edx_video_id)
```
`encoded_videos` or `encoded_video`?

noraiz-anwar commented:
jenkins run bokchoy

awaisdar001 commented:
@noraiz-anwar this looks good to me, the only thing which i think is missing is a sandbox on your branch. Please get the help within the team to have it.

Thanks.

noraiz-anwar commented:
@awaisdar001 yeah sure.
Thanks for you time.

noraiz-anwar commented on lms/djangoapps/courseware/tests/test_video_mongo.py:
```diff
@@ -583,30 +583,97 @@ def test_get_html_with_mocked_edx_video_id(self):
     def test_get_html_with_existing_edx_video_id(self):
-        # create test profiles and their encodings
+        """
+        Tests the `VideoModule` `get_html` where `edx_video_id` is given and related video is found
```
`VideoModule` is a module name. According to recommendations stated [here](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python_guidelines.html), this is how we are suppose to write.
Thanks

noraiz-anwar commented:
@awaisdar001 Sandbox is created. Please have a look.

awaisdar001 commented:
@noraiz-anwar Sandbox is ready but I can not test the behaviour there by adding a tab.
I can not see video uploads option also. 
https://github.com/edx/edx-platform/blob/6d2468d0dcf782a7561ecc61946146d6fae2762f/cms/templates/widgets/header.html#L70-L73

Is this significant amount of work just to get it test on the sandbox?

noraiz-anwar commented:
Here is the actual [block](https://studio-tnl-5769.sandbox.edx.org/container/block-v1:arbisoft+CS100+2016_t1+type@vertical+block@84bf1bea2d2c4d80b84ac7008f107fd0) on which issue was reported. You can test the behavior after adding tab to Video ID of this block or any other.
I have uploaded that course (mentioned in issue) on sandbox.

awaisdar001 commented:
just a nit looks good to me otherwise.

awaisdar001 commented:
👍 after addressing the nit and tests are passed.

noraiz-anwar commented:
@awaisdar001 yes, for sure.

noraiz-anwar commented:
jenkins run lettuce

noraiz-anwar commented:
jenkins run bokchoy
